cpp/src/binaryop/compiled/operation.cuh

```diff
@@ -91,11 +91,46 @@ struct TrueDiv {
 };
 
 struct FloorDiv {
-  template <typename T1, typename T2>
-  __device__ inline auto operator()(T1 const& lhs, T2 const& rhs)
-    -> decltype(floor(static_cast<double>(lhs) / static_cast<double>(rhs)))
+  template <typename TypeLhs,
+            typename TypeRhs,
+            std::enable_if_t<(std::is_integral_v<std::common_type_t<TypeLhs, TypeRhs>> and
+                              std::is_signed_v<std::common_type_t<TypeLhs, TypeRhs>>)>* = nullptr>
+  __device__ inline auto operator()(TypeLhs x, TypeRhs y) -> decltype(x / y)
+  {
+    if ((x ^ y) >= 0) {
+      return x / y;
+    } else {
+      auto const quotient  = x / y;
+      auto const remainder = x % y;
+      return quotient - !!remainder;
+    }
+  }
```

Comment:
A branch-free implementation looks like it performs better across the board:

```
gbench/tools/compare.py -a --no-utest benchmarks cudf/cpp/build/benchmarks/branchy.json cudf/cpp/build/benchmarks/nonbranchy.json  
Comparing cudf/cpp/build/benchmarks/branchy.json to cudf/cpp/build/benchmarks/nonbranchy.json
Benchmark                                                                                          Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/10000/manual_time_mean                      -0.0282         -0.0244            43            41            63            62
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/10000/manual_time_median                    -0.0286         -0.0249            43            41            63            62
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/10000/manual_time_stddev                    +0.2685         +0.4944             0             0             0             0
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/100000/manual_time_mean                     -0.0365         -0.0300            46            44            66            64
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/100000/manual_time_median                   -0.0365         -0.0300            46            44            66            64
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/100000/manual_time_stddev                   -0.0921         -0.1937             0             0             0             0
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/1000000/manual_time_mean                    -0.0682         -0.0562            80            74           100            95
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/1000000/manual_time_median                  -0.0683         -0.0562            80            74           100            95
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/1000000/manual_time_stddev                  -0.0394         -0.1046             0             0             0             0
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/10000000/manual_time_mean                   -0.0862         -0.0828           425           388           446           410
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/10000000/manual_time_median                 -0.0862         -0.0827           425           388           446           410
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/10000000/manual_time_stddev                 +0.7530         -0.0782             0             0             0             0
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/100000000/manual_time_mean                  -0.1000         -0.1000          3722          3350          3763          3387
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/100000000/manual_time_median                -0.1000         -0.1000          3722          3350          3763          3386
COMPILED_BINARYOP/FLOOR_DIV_int64_t_int64_t_int64_t/100000000/manual_time_stddev                +0.0632         +0.1715             0             0             1             1
```

Reply:
Thread divergence could easily be the number one performance killer for such a small kernel. The new branchless algorithm looks great.

Reply:
What are you comparing against? The old version doesn't appear to have branches either.

Reply:
The previous implementation on this branch: 

https://github.com/rapidsai/cudf/pull/12131/commits/bd5bc8895c4256760b06fe6778485bddab4388e9#diff-783a8c11b115684b22107cfc1316116a7ea17b6e5df50042f31627ff1869d34eR102